<a href="https://colab.research.google.com/github/Deepanshuduttpant/wallmart-sale/blob/main/Wallmart_sales_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd 
import numpy as np
import plotly.express as px

In [ ]:
df1 = pd.read_csv('/content/Walmart.csv')

# Basic level analysis

In [ ]:
df1.head() 

Order ID  Order Date   Ship Date    Customer Name        Country  \
0  CA-2013-138688  13-06-2013  17-06-2013  Darrin Van Huff  United States   
1  CA-2011-115812  09-06-2011  14-06-2011  Brosina Hoffman  United States   
2  CA-2011-115812  09-06-2011  14-06-2011  Brosina Hoffman  United States   
3  CA-2011-115812  09-06-2011  14-06-2011  Brosina Hoffman  United States   
4  CA-2011-115812  09-06-2011  14-06-2011  Brosina Hoffman  United States   

          City       State     Category  \
0  Los Angeles  California       Labels   
1  Los Angeles  California  Furnishings   
2  Los Angeles  California          Art   
3  Los Angeles  California       Phones   
4  Los Angeles  California      Binders   

                                        Product Name   Sales  Quantity  Profit  
0  Self-Adhesive Address Labels for Typewriters b...   14.62       2.0    6.87  
1  Eldon Expressions Wood and Plastic Desk Access...   48.86       7.0   14.17  
2                                         Newell 322    7.28       4.0    1.97  
3                     Mitel 5320 IP Phone VoIP phone  907.15       4.0   90.72  
4  DXL Angle-View Binders with Locking Rings by S...   18.50       3.0    5.78

to see the total no of rows and columns

In [ ]:
df1.shape

(3203, 12)

to see the names of the columns

In [ ]:
list(df1.columns)

['Order ID',
 'Order Date',
 'Ship Date',
 'Customer Name',
 'Country',
 'City',
 'State',
 'Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Profit']

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3203 entries, 0 to 3202
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order ID       3203 non-null   object 
 1   Order Date     3203 non-null   object 
 2   Ship Date      3203 non-null   object 
 3   Customer Name  3203 non-null   object 
 4   Country        3203 non-null   object 
 5   City           3203 non-null   object 
 6   State          3203 non-null   object 
 7   Category       3203 non-null   object 
 8   Product Name   3203 non-null   object 
 9   Sales          3203 non-null   float64
 10  Quantity       3203 non-null   float64
 11  Profit         3203 non-null   float64
dtypes: float64(3), object(9)
memory usage: 300.4+ KB


to see the null values 

In [ ]:
df1.isna().sum()

Order ID         0
Order Date       0
Ship Date        0
Customer Name    0
Country          0
City             0
State            0
Category         0
Product Name     0
Sales            0
Quantity         0
Profit           0
dtype: int64

In [ ]:
df1.nunique()

Order ID         1611
Order Date        845
Ship Date         911
Customer Name     686
Country             1
City              169
State              11
Category           17
Product Name     1494
Sales            2295
Quantity           14
Profit           2164
dtype: int64

converting order date and ship date datatype to datetime 

In [ ]:
df1['Order Date'] = pd.to_datetime(df1['Order Date'])
df1['Ship Date'] = pd.to_datetime(df1['Ship Date'])

<ipython-input-145-a7fea54ccc0d>:1: UserWarning:

Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.

<ipython-input-145-a7fea54ccc0d>:2: UserWarning:

Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.



Extracting month and year column from order date

In [ ]:
df1['Month'] = df1['Order Date'].dt.month
df1['Year'] = df1['Order Date'].dt.year

Adding some more columns in our DataFrame 

In [ ]:
df1['Cost'] = df1['Sales']-df1['Profit']

In [ ]:
df1['Cost/Unit'] = df1['Cost']/df1['Quantity']
df1['Price'] = df1['Sales']/df1['Quantity']
df1['Profit/Unit'] = df1['Profit']/df1['Quantity']

In [ ]:
df1.head(1)

Order ID Order Date  Ship Date    Customer Name        Country  \
0  CA-2013-138688 2013-06-13 2013-06-17  Darrin Van Huff  United States   

          City       State Category  \
0  Los Angeles  California   Labels   

                                        Product Name  Sales  Quantity  Profit  \
0  Self-Adhesive Address Labels for Typewriters b...  14.62       2.0    6.87   

   Month  Year  Cost  Cost/Unit  Price  Profit/Unit  
0      6  2013  7.75      3.875   7.31        3.435

setting the columns in our dataframe according to our convenience

In [ ]:
 df = df1.reindex(columns = ['Order ID',	'Order Date',	'Month',	'Year'	,'Ship Date',	'Customer Name',	'Country',	'City',	'State',	'Category',	'Product Name','Quantity','Cost','Sales','Profit','Cost/Unit','Price','Profit/Unit'])

In [ ]:
df.head(2)

Order ID Order Date  Month  Year  Ship Date    Customer Name  \
0  CA-2013-138688 2013-06-13      6  2013 2013-06-17  Darrin Van Huff   
1  CA-2011-115812 2011-09-06      9  2011 2011-06-14  Brosina Hoffman   

         Country         City       State     Category  \
0  United States  Los Angeles  California       Labels   
1  United States  Los Angeles  California  Furnishings   

                                        Product Name  Quantity   Cost  Sales  \
0  Self-Adhesive Address Labels for Typewriters b...       2.0   7.75  14.62   
1  Eldon Expressions Wood and Plastic Desk Access...       7.0  34.69  48.86   

   Profit  Cost/Unit  Price  Profit/Unit  
0    6.87   3.875000   7.31     3.435000  
1   14.17   4.955714   6.98     2.024286

finding no. of years present in our dataframe

In [ ]:
#df.Year.value_counts().sort_values(by='Year')
zzz=df['Year'].value_counts().sort_index(ascending=False)
#'''we cant write here sort_values becoz if we write then becoz value_count() k baad sort_value() aaega islie
#ye value counts() k column ko sort krega in decending order. but when u write sort_index() then ye
#value_counts() ko sort na krke index k behalf pe sort kkrega'''  

# '''imp note single column me "by" argument use nhi hota with sort_values()...it only uses with 2 column 
# names, or groupby func...or merge func()...'''
zzz


2014    1099
2013     801
2012     642
2011     661
Name: Year, dtype: int64

**year wise finding-- monthly= (profit and sales)**

In [ ]:
year1=df[df.Year==2011]
year1.shape

(661, 18)

**Now- month wise showing the profit and sales of every state** . 

In [ ]:
# after applying merge function it already gives your answer into dataframe you dont have to convert it
# into dataframe
q2 = year1.groupby(['Month','State'])['Profit'].sum()
c2 = year1.groupby(['Month','State'])['Sales'].sum()
d2 = pd.merge(q2,c2,how='left',on=['Month','State'])
d2.reset_index(inplace=True)# We use the inplace=True argument when we want to modify a DataFrame object in place, without creating a new DataFrame object.
d2.sort_values(by='Sales', ascending = True)
d2
# after this code i can give you examples that whats d/f b/w reset_index() and reset_index(inplace=True)

Month       State   Profit     Sales
0       1     Arizona  -288.01    378.59
1       1  California   380.59   2680.13
2       1    Colorado   252.32   2027.15
3       1      Nevada    19.24     44.79
4       1      Oregon   -33.25    664.76
..    ...         ...      ...       ...
66     12  California  2273.83  20722.05
67     12    Colorado   110.47   1303.21
68     12      Nevada    59.49    475.94
69     12  New Mexico     2.30     57.27
70     12  Washington    32.55    165.45

[71 rows x 4 columns]

In [ ]:
#Example 1: Using reset_index()
# create a sample DataFrame
df = pd.DataFrame({'Name': ['Alice', 'Bob', 'Charlie'], 'Score': [90, 85, 95]})

# reset the index without using inplace=True
df_new = df.reset_index()

# print the original DataFrame and the new DataFrame
print(df)
print(df_new)


      Name  Score
0    Alice     90
1      Bob     85
2  Charlie     95
   index     Name  Score
0      0    Alice     90
1      1      Bob     85
2      2  Charlie     95


In [ ]:
#Example 2: Using reset_index(inplace=True)
# create a sample DataFrame
df = pd.DataFrame({'Name': ['Alice', 'Bob', 'Charlie'], 'Score': [90, 85, 95]})

# reset the index using inplace=True
df.reset_index(inplace=True)

# print the modified DataFrame
print(df)
# In this example, reset_index(inplace=True) is used to reset the index of the DataFrame df in place, without creating a new DataFrame.


   index     Name  Score
0      0    Alice     90
1      1      Bob     85
2      2  Charlie     95


In [ ]:
''' Q - agar merge krte samai hm on = khali state column pe lagaye or month column add na krre then
it gives me 635 rows but agar hm month lgaye then it gives 71 rows only jbki jo month and state q2
me hai voohi c2 me b hai. to same aana chahiye then why it gives me different values.'''

q2 = year1.groupby(['Month','State'])['Profit'].sum()
c2 = year1.groupby(['Month','State'])['Sales'].sum()
d2 = pd.merge(q2,c2,how='left',on=['Month','State'])
d2.reset_index(inplace=True)
d2

Month       State   Profit     Sales
0       1     Arizona  -288.01    378.59
1       1  California   380.59   2680.13
2       1    Colorado   252.32   2027.15
3       1      Nevada    19.24     44.79
4       1      Oregon   -33.25    664.76
..    ...         ...      ...       ...
66     12  California  2273.83  20722.05
67     12    Colorado   110.47   1303.21
68     12      Nevada    59.49    475.94
69     12  New Mexico     2.30     57.27
70     12  Washington    32.55    165.45

[71 rows x 4 columns]

making a line graph by ploty

In [ ]:
#import ploty.express as px
fig = px.line(d2,'Month','Profit',color='State',title='Profit over Months(2011)')
fig.show()

CALIFORNIA profit inc. over month and its highest profit is in month 12, i.e - 2273

In [ ]:
# graph of sales over month
fig = px.line(d2,y='Sales',x='Month',color='State',title = 'Sales_over_month(2011)')
fig.show()

california sales inc. over months and its highest sale is on month 12 which is 20722.

In [ ]:
# Q- agar ab hame isme ye nikalna hai ki 12th month me new mexico state ki kitni sale hui hai
c= year1[year1['State']=='New Mexico']
c

Order ID Order Date  Month  Year  Ship Date    Customer Name  \
1114  CA-2011-143385 2011-08-31      8  2011 2011-05-09  Shirley Jackson   
1356  CA-2011-152345 2011-12-29     12  2011 2012-03-01         Shui Tom   
1357  CA-2011-152345 2011-12-29     12  2011 2012-03-01         Shui Tom   
2733  CA-2011-100881 2011-03-28      3  2011 2011-01-04    Daniel Raglin   
2976  US-2011-134054 2011-10-10     10  2011 2011-10-10       Fred Chung   

            Country         City       State     Category  \
1114  United States     Santa Fe  New Mexico  Accessories   
1356  United States  Albuquerque  New Mexico       Phones   
1357  United States  Albuquerque  New Mexico      Storage   
2733  United States  Albuquerque  New Mexico       Phones   
2976  United States  Albuquerque  New Mexico          Art   

                                           Product Name  Quantity    Cost  \
1114               KeyTronic KT400U2 - Keyboard - Black       9.0   74.02   
1356                Mediabridge Sport Armband iPhone 5s       3.0   29.67   
1357  Fellowes Bases and Tops For Staxonsteel/High-S...       1.0   25.30   
2733                                       AT&T TR1909W       3.0  279.70   
2976  Dixon Ticonderoga Core-Lock Colored Pencils, 4...       7.0  143.28   

       Sales  Profit  Cost/Unit       Price  Profit/Unit  
1114   92.52   18.50   8.224444   10.280000     2.055556  
1356   23.98   -5.69   9.890000    7.993333    -1.896667  
1357   33.29    7.99  25.300000   33.290000     7.990000  
2733  302.38   22.68  93.233333  100.793333     7.560000  
2976  255.85  112.57  20.468571   36.550000    16.081429

In [ ]:
# ''' IMP CODE HAI LIKH LENA'''----- .query
result = year1.query('Month == 12 and State == "New Mexico"')['Sales']
result

1356    23.98
1357    33.29
Name: Sales, dtype: float64

In [ ]:
# or u can apply this also
c = year1[(year1['State']=='New Mexico') & (year1['Month']==12)]
sale_data = c['Sales'].sum()
c

Order ID Order Date  Month  Year  Ship Date Customer Name  \
1356  CA-2011-152345 2011-12-29     12  2011 2012-03-01      Shui Tom   
1357  CA-2011-152345 2011-12-29     12  2011 2012-03-01      Shui Tom   

            Country         City       State Category  \
1356  United States  Albuquerque  New Mexico   Phones   
1357  United States  Albuquerque  New Mexico  Storage   

                                           Product Name  Quantity   Cost  \
1356                Mediabridge Sport Armband iPhone 5s       3.0  29.67   
1357  Fellowes Bases and Tops For Staxonsteel/High-S...       1.0  25.30   

      Sales  Profit  Cost/Unit      Price  Profit/Unit  
1356  23.98   -5.69       9.89   7.993333    -1.896667  
1357  33.29    7.99      25.30  33.290000     7.990000

 which state gives you highest profit in the whole year

In [ ]:

state_profit = year1.groupby('State')['Profit'].sum()
state_profit = state_profit.sort_values(ascending=False).reset_index()
state_profit

State    Profit
0  California  12638.00
1  Washington   6607.25
2        Utah    822.67
3    Colorado    438.21
4      Nevada    250.27
5  New Mexico    156.05
6     Montana    129.84
7       Idaho     62.05
8      Oregon   -135.41
9     Arizona   -903.19

In [ ]:
#fig = px.line(state_profit,x='State',y='Profit')
#fig.show()
state_profit = year1.groupby('State')['Profit'].sum().reset_index()
fig = px.line(state_profit,x='State',y='Profit',title = 'Yearly profit per state (2011)')
fig.show()

'''in this code aapko reset_index lagana jaruri hoga chahe to ye state_profit ki line dubara likh
k isme .reset_index() lagaye ya phir iske upper vaale code me sort_values k saath reset index lagaye...
["profit.sum() vaali line (1st line) me mat lga dena" bcoz reset_index() code k last me lgta hai,
or 1st line k baad b 2nd line me state_profit me hum sort func. ka use krre hai...islie vo 2nd line me
aaega...]'''

'''ab ye reset_index lagana jaruri isliye hai becoz reset_index() lagane se ye full data ko ek data
frame me convert kr deta hai....or plotly dataframe vaali chijo ka hi graph banata hai.'''


'ab ye reset_index lagana jaruri isliye hai becoz reset_index() lagane se ye full data ko ek data\nframe me convert kr deta hai....or plotly dataframe vaali chijo ka hi graph banata hai.'

CALIFORNIA is he most profitable state 
and
 ARIZONA is the least profitable state.

 Which state gives most sales to wallmart in 2011 ?

In [ ]:
state_sales = year1.groupby('State')['Sales'].sum()
state_sales = state_sales.sort_values(ascending = False).reset_index()
state_sales

State     Sales
0  California  91303.54
1  Washington  29871.57
2     Arizona   8295.25
3    Colorado   6502.29
4      Oregon   4807.81
5        Utah   2943.69
6      Nevada   2561.20
7  New Mexico    708.02
8       Idaho    465.21
9     Montana    424.49

In [ ]:
state_sales = year1.groupby('State')['Sales'].sum().reset_index()
fig = px.line(state_sales,x='State',y='Sales',title = 'Yearly SALES per State(2011)')
fig.show()

'CALIFORNIA' has highest and 'MONTANA' has lowest SALE in year 2011

In [ ]:
xcz = df.Year.value_counts().sort_index()
''' sort_values() kha lgna hai and sort_index() kha pe iska d/f smjh aagya na'''
#sort_index is used to sort series and sort values is used to sort DataFrame
xcz

In [ ]:
year2 = df[df['Year']==2012]
year2.shape

In [ ]:
year2.head()

Q-1 find profit and sales over months of every state in 2012 

Q-2 find total (profit and sales) over states in year 2012

In [ ]:
a1 = year2.groupby(['Month','State'])['Profit'].sum()
a2 = year2.groupby(['Month','State'])['Sales'].sum()
a3 = pd.merge(a1,a2,how='left',on=['Month','State']).reset_index()
a3

Q- which state gives most of the profit over time.

In [ ]:
fig = px.line(a3,x='Month',y='Profit',color='State',title='profit over months(2012)')
fig.show()

CALIFORNIA gives most of the profit over time

In [ ]:
fig = px.line(a3,y='Sales',x='Month',color='State',title='Sales over months(2012)')
fig.show()

CALIFORNIA gives max. Sales over time.

Q- Yearly profit of each state.

In [ ]:
state_profit2 = year2.groupby('State')['Profit'].sum()
state_profit2 = state_profit2.sort_values(ascending = False).reset_index()
state_profit2
    


In [ ]:
fig = px.line(state_profit2,x='State',y='Profit',title='Yearly profit per states(2012)')
fig.show()
# this graph is made by above state_profit2 vha pe humne sort_values kr rkhi thi profit me islie ye graph
# aisa aaya or upper humne dubara code nhi likha islie aisa hua

In [ ]:
state_profit2 = year2.groupby('State')['Profit'].sum().reset_index()
fig = px.line(state_profit2,x='State',y='Profit',title='Yearly profit per states(2012)')
fig.show()
# is graph me ye state_profit2 hai vo saamne hi likh dia or isme sorting nhi lgi hai islie values 
# aache se ai hai.


CALIFORNIA made highest profit of 14371

Q- Yearly Sales of each state

In [ ]:
state_sales2 = year2.groupby('State')['Sales'].sum()
state_sales2 = state_sales2.sort_values(ascending=False).reset_index()
state_sales2

In [ ]:
state_sales2 = year2.groupby('State')['Sales'].sum().reset_index()
fig = px.line(state_sales2,x='State', y='Sales',title='Yearly sales per states(2012)')
fig.show()

CALIFORNIA make a highest sale of 88443

Q-1 find profit and sales over months of every state in 2013

In [ ]:
year3 = df[df['Year']==2013]
year3.shape

In [ ]:
b1 = year3.groupby(['Month','State'])['Profit'].sum()
b2 = year3.groupby(['Month','State'])['Sales'].sum()
b3 = pd.merge(b1,b2,how='left',on=['Month','State']).reset_index()
b3

In [ ]:
fig = px.line(b3,x='Month',y='Profit',color='State',title='Profit over months(2013)')
fig.show()

In [ ]:
fig = px.line(b3,x='Month',y='Sales',color='State',title='Sales over Months(2013)')
fig.show()

finding Total sales and total profit in year 2013

In [ ]:
state_profit3 = year3.groupby('State')['Profit'].sum()
state_profit3

In [ ]:
state_profit3 = year3.groupby('State')['Profit'].sum().reset_index()
fig= px.line(state_profit3,x='State',y='Profit',title='Total Profit Over States(2013)')
fig.show()

In [ ]:
state_sales3 = year3.groupby('State')['Sales'].sum()
state_sales3

In [ ]:
state_sales3 = year3.groupby('State')['Sales'].sum().reset_index()
fig = px.line(state_sales3,x='State',y='Sales',title='total sales over states(2013)')
fig.show()

Now for year 2014

Find Sales and Profit of every states over months

In [ ]:
year4 = df[df['Year']==2014]

In [ ]:
f1 = year4.groupby(['Month','State'])['Profit'].sum()
f2 = year4.groupby(['Month','State'])['Sales'].sum()
f3 = pd.merge(f1,f2,how='left',on=['Month','State']).reset_index()
f3

In [ ]:
fig = px.line(f3,x='Month',y='Profit',color = 'State',title='Profit Over Months(2014)')
fig.show()

In [ ]:
fig = px.line(f3,x='Month',y='Sales',color = 'State',title='Sales Over Months(2014)')
fig.show()

Q- find total profit and *sales* over states of year 2014

In [ ]:
state_profit4 = year4.groupby('State')['Profit'].sum()
state_profit4 = state_profit4.sort_values(ascending=False)
state_profit4

In [ ]:
state_profit4 = year4.groupby('State')['Profit'].sum().reset_index()
fig = px.line(state_profit4,x='State',y='Profit',title = 'Total Profit over states(2014)')
fig.show()

In [ ]:
state_sales4 = year4.groupby('State')['Sales'].sum()
state_sales4 = state_sales4.sort_values(ascending = False)
state_sales4


In [ ]:
state_sales4 = year4.groupby('State')['Sales'].sum().reset_index()

In [ ]:
fig = px.line(state_sales4,x='State',y='Sales',title = 'Total Profit over states(2014)')
fig.show()

Overall profit and sales of wallmart store over months

In [ ]:
z1 = df.groupby(['Month','State'])['Profit'].sum()
z2 = df.groupby(['Month','State'])['Sales'].sum()
z3 = pd.merge(z1,z2,how='left',on=['Month','State']).reset_index()
z3

In [ ]:
fig = px.line(z3,x='Month',y='Profit',color='State',title = '(Overall) Profit over Months')
fig.show()

In [ ]:
fig = px.line(z3,x='Month',y='Sales',color='State',title = '(Overall) Sales over months')
fig.show()

Total Sales and Profit of Whole wallmart data

In [ ]:
total_state_profit = df.groupby('State')['Profit'].sum()
total_state_profit = total_state_profit.sort_values(ascending= False)
total_state_profit 

In [ ]:
total_state_profit = df.groupby('State')['Profit'].sum().reset_index()
fig = px.line(total_state_profit,x='State', y='Profit',title = '(Overall) profit over states')
fig.show()

In [ ]:
total_state_sales = df.groupby('State')['Profit'].sum()
total_state_sales = total_state_sales.sort_values(ascending= False)
total_state_sales

In [ ]:
total_state_sales = df.groupby('State')['Sales'].sum().reset_index()
fig = px.line(total_state_sales,x='State', y='Sales',title = '(Overall) Sales over States')
fig.show()



Overall conclusion is that - california has highest no. of profit and sales in all over the years and in total years


Now generating category wise sales and profit

In [ ]:
df.head()

Q-- How many times Sales occured from each city. 

In [ ]:
cities = df.City.value_counts().reset_index()
cities
cities.rename(columns={'index':'City','City':'Count'})

How many times Sales occured from each State

In [ ]:
df.State.value_counts().reset_index()


How many times Sales occured from each category

In [ ]:
df.Category.value_counts().reset_index()

In [ ]:
z1 = df.groupby('Category')['Profit'].sum().reset_index()
z1=z1.sort_values(by='Profit',ascending=False)
z1
#'''  isme "reset_index() lgaya hua hai islie ye ek dataframe me convert ho gya hai or datafrme me 
#sort_values() me "by" argument lagana pdta hai....vhi agar hm isse data frame me convert nhi krte to
#"by" argument lagane pr ye hame "error" dikha deta...
#ab data frame me convert krne pr hame sort_index() ki b tension nhi hogi kyunki ye humse poochta hai
#ki "by= 'kis column ko sort krna cha rhe ho' '''

In [ ]:
z1 = df.groupby('Category')['Profit'].sum()
z1=z1.sort_values(ascending=False)
z1
# '''  ab kyunki pehli vaale code me .reset_index() upper tha uske baad sorting hui islie index no.
#line se nhi aaye...or agar reset_index() niche likhoge to grupby krne pe dataframe nhi bnega pehli
#line me islie hame "by" argument hatana pdega or is case me sort_index() b kaam aa skta hai.

# just see the d/f b/w both of the code...

In [ ]:
z1 = df.groupby('Category')['Profit'].sum().reset_index()
fig = px.line(z1,x='Category',y='Profit',title='Profit over categories')
fig.show()

COPIERS  has the highest profit of 19327

In [ ]:
y1 = df.groupby('Category')['Sales'].sum()
y1

In [ ]:
y1=y1.reset_index()
fig = px.line(y1,x='Category',y='Sales',title = 'Sales over Caegories')
fig.show()

CHAIRS has highest sales of - 101781

In [ ]:
df.head()

In [ ]:
# showing above graph by bar graph
y1=y1.reset_index()
#fig=px.bar(y1,x='State',y='Sale',color='State',title='Profit By Categroy')
fig = px.bar(y1, x='Category', y='Sales', color='Category', title='Sales By Category',hover_name= 'Category')
fig.show()

In [ ]:
z1 = df.groupby('Category')['Profit'].sum().reset_index()
fig= px.bar(z1,x='Category',y='Profit',color ='Category',title='Profit Over Category',hover_name = 'Category')
fig.show()

Total profit by state and city in dataframe df

In [ ]:
c = df.groupby(['State','City'])['Profit'].sum()
c

Total sales by State and city of DataFrame df

In [ ]:
d = df.groupby(['State','City'])['Sales'].sum()
d

In [ ]:
e = df.groupby(['State','City'])['Quantity'].mean()
e
# Q - why we extract mean of this data or mean nikalne se kya faida hota hai...

State       City         
Arizona     Avondale         5.000000
            Bullhead City    1.500000
            Chandler         3.285714
            Gilbert          3.866667
            Glendale         2.913043
                               ...   
Washington  Renton           4.666667
            Seattle          3.714953
            Spokane          5.142857
            Vancouver        3.000000
Wyoming     Cheyenne         4.000000
Name: Quantity, Length: 170, dtype: float64

No. of unique customers present in each city. 

In [ ]:
f=df.groupby(['State','City'])['Customer Name'].nunique()
f

State       City         
Arizona     Avondale           3
            Bullhead City      1
            Chandler           6
            Gilbert            5
            Glendale          13
                            ... 
Washington  Renton             1
            Seattle          188
            Spokane            5
            Vancouver          3
Wyoming     Cheyenne           1
Name: Customer Name, Length: 170, dtype: int64

Name of the customers present in each cities of d/f States 

In [ ]:
#w = df.groupby(['State','City'])['Customer Name'].value_counts()
#w
# when applying above code then it gives you answer in series, and when u apply .reset_index then it gives u value_error. there4 u have to apply 2nd code or below code. --- appy both the code then you will see the d/f. 
w = df.groupby(['State','City'])['Customer Name'].value_counts().reset_index(name='Count').reset_index(drop=True)
w
# yha se aap states wise no. of orders bhi nikal skte hai.

State           City    Customer Name  Count
0        Arizona       Avondale     Kunst Miller      3
1        Arizona       Avondale     Philip Brown      2
2        Arizona       Avondale          Amy Cox      1
3        Arizona  Bullhead City    Gary Zandusky      2
4        Arizona       Chandler     James Lanier      2
...          ...            ...              ...    ...
1459  Washington        Spokane     Meg O'Connel      1
1460  Washington      Vancouver   Anthony Rawles      3
1461  Washington      Vancouver     Frank Merwin      1
1462  Washington      Vancouver        Jane Waco      1
1463     Wyoming       Cheyenne  Michelle Arnett      1

[1464 rows x 4 columns]

In [ ]:
# now i want to find the no. of orders of each city like in avondale city no.of orders = 6
w = df.groupby(['State','City'])['Customer Name'].value_counts()
city_counts = w.groupby(level=[0, 1]).sum() # The groupby method with level=[0, 1] specifies that we want to group by the first and second levels of the index (State and City) and then apply the sum method to get the sum of the value_counts for each City.
#here (level=[0,1]).sum()-- means level-0 = 0th column of w and level-1 = 1st column and then sum it.
# bole to = ex- jha jha state = arizona hai(level-0) and jha khi bhi city=avondale ho unka sum krdo.
print(city_counts)



State       City         
Arizona     Avondale           6
            Bullhead City      2
            Chandler           7
            Gilbert           15
            Glendale          23
                            ... 
Washington  Renton             3
            Seattle          428
            Spokane            7
            Vancouver          5
Wyoming     Cheyenne           1
Name: Customer Name, Length: 170, dtype: int64


In [ ]:
# it also gives us the same answer, as above code gives.
s1= df.groupby(['State', 'City']).size() # The size() method in Pandas is used to get the 
#number of rows (i.e., observations) in each group of a groupby object.
# means ex- state = arizona , and city = avondale, jha b ye dikhe unko grp by krdo or unka size(jitni bar ye aae hai means count hue hai usko batao)
print(s1)

''' when you run c,d,e,f,w then you will find there is name ,length,dtype is - written in last line but in s1 or in this code, when you run this then- Name will not come in last, so bcoz of this when you merge s1 with any other line it will gives you a value error- ValueError: Cannot merge a Series without a name. so to resolve this error you have to  assign a name to s1
assigning a name to s1 by two methods = 1st method.-  in second line write- s1.name = 'Size' (or any name u can give)
2nd method- s1 = df.groupby(['State', 'City']).size().rename('Size')'''



State       City         
Arizona     Avondale           6
            Bullhead City      2
            Chandler           7
            Gilbert           15
            Glendale          23
                            ... 
Washington  Renton             3
            Seattle          428
            Spokane            7
            Vancouver          5
Wyoming     Cheyenne           1
Length: 170, dtype: int64


" when you run c,d,e,f,w then you will find there is name ,length,dtype is - written in last line but in s1 or in this code, when you run this then- Name will not come in last, so bcoz of this when you merge s1 with any other line it will gives you a value error- ValueError: Cannot merge a Series without a name. so to resolve this error you have to  assign a name to s1\nassigning a name to s1 by two methods = 1st method.-  in second line write- s1.name = 'Size' (or any name u can give)\n2nd method- s1 = df.groupby(['State', 'City']).size().rename('Size')"

In [ ]:
m= pd.merge(c,d,how='left',on=['State','City'])
m=pd.merge(pd.merge(m,e,how='left',on=['State','City']),f,how='left',on=['State','City'])
m= pd.merge(m,s1 ,how='left',on=['State','City'])
m.rename({'Quantity':'Avg. Order Quantity', 'Customer Name_y':'no._of_orders',
'Customer Name_x':'Customers',},axis=1,inplace=True)
m
#m.rename(columns={'Quantity':'Avg. Order Quantity'})
#isme column ko rename krne k lie rename method ka use kia tha but in that code we use-- (g.column=['value_counts'])

ValueError: ignored

In [ ]:
xyz = m.reset_index()
xyz

In [ ]:
h= df.groupby(['State','City'])['Customer Name'].nunique()
h

In [ ]:
df

In [ ]:
f = df.groupby(['State','City'])['Customer Name'].nunique()
print(f)

In [ ]:
g = df.groupby(['State','City'])['Customer Name'].value_counts()
print(g)

In [ ]:
f_df = pd.DataFrame(f)
f_df

In [ ]:
g = pd.DataFrame(g)
g

In [ ]:
#f_df.columns = ['City']  # Rename the column to 'Count' instead of 'Customer Name'
f_df=f_df.rename(columns={'Customer Name':'value_counts'})
f_df= f_df.reset_index()
#'''g[g['City']=='Avondale']'''


In [ ]:
#g.rename({'Customer Name' : 'value' },axis=1,inplace=True)#you can also use this line to change column name
#g.columns = ['value_counts']  # Rename the column to 'Count' instead of 'Customer Name'
g=g.rename(columns={'Customer Name':'value_counts'})
g = g.reset_index()

In [ ]:
#  Q- to ask -isko run krne k lie mjhe pehle g.reset_index() and g.count kyu lagaya 
f_df[f_df['State']=='Arizona']


In [ ]:
g[g['State']=='Arizona']

In [ ]:
f_df = f_df.reset_index().drop('index', axis=1)
f_df.columns = ['State', 'City', 'Count']


In [ ]:
# in do (f_df and g) examples se ye conclusion nikal k aata hai ki no dataframe have same_name 2 columns
# g me dataframe bnane k baad jb if u see the dataframe it have 2 columns name same- i.e Customer Name
# or koi bhi 2 columns k same naam nhi ho skte islie we have to change 1 columns name

Q- Is there any relation between Sales & Profit

In [ ]:
 a = df.groupby(['State'])['Profit'].sum()

In [ ]:
b=df.groupby(['State'])['Sales'].sum()

In [ ]:
xy = pd.merge(a,b,how='left',on=['State']).reset_index()
xy.sort_values(by='Profit')
xy
# imp-- jb bhi aapko random index dikhe rows ka(jaise is code me dikh rha hai) or aap usko hta k reset index lgana chahte ho to bs agli line me reset_index(Drop=True) lga dena to ye random index value ko hta k reset index lga dega. jaise mene agle code me lgake dikhaya hai.    dono code ko run krke dekhna u can see the d/f.

In [ ]:
ab = pd.merge(a,b,how='left',on=['State']).reset_index()
ab.sort_values(by = 'Profit').reset_index(drop=True)
ab

**Graph of Sales and Profit Co-Relation by State**

In [ ]:
fig = px.scatter(ab,x='Sales',y='Profit',color='State',title='Sales and Profit Co-Relation by State',size='Sales',hover_name='State')
fig.show()

# In the given code, size='Profit' is used to set the size of the data points in the scatter plot based on the value of the 'Profit' column. This means that data points with higher 'Profit' values will be represented by larger markers in the plot, while data points with lower 'Profit' values will be represented by smaller markers.

In [ ]:
#this graph is the improve version of above graph
#by this line - '''fig.update_traces(marker=dict(size=10))''' you can change the size of traces in graph
fig = px.scatter(ab, x='Sales', y='Profit', color='State', 
                 title='Sales and Profit Co-Relation by State', 
                 hover_name='State', size_max=10)
fig.update_traces(marker=dict(size=10))
fig.show()



Graph of Sales & Profit Co-Relation by City

In [ ]:
fig=px.scatter(ab,'Sales','Profit',color='State',title='Sales & Profit Co-Relation by City',
               size='Sales', hover_name='City')
fig.show()
# now see how hover_name is imp. , jb hover_name = City daala to jitni cities hai utne point plot honge
# it means--(hover_name) points plot krne k kaam aata hai and (color = state) jo right side me  states name
# mark hote hai uske kaam aata hai...(by seeing these both chart you can see the diff.)

In [ ]:
# agar hum pichle code ki jagah ye code likhenge na to error dikhaega
fig=px.scatter(ab,'Profit','Sales',color='State',title='Sales & Profit Co-Relation by City',
               size='Profit', hover_name='City')
fig.show()
# becoz humne (size = 'Profit') likha hai, means jitna bada profit utna hi bada gole ka size
# but agar profit -ve me hua to---tb gola to dikhega hi nhi(or kyunki hamara profit -ve me hai)...islie size = profit pe error dikhaega

 Q--Is there any relation between No. of Customers & Sales?

In [ ]:
zx = df.groupby(['State'])['Customer Name'].nunique().reset_index() # .reset_index() yha islie lga bcoz in next line we r renaming the column and rename func. only works in datafrme, or jb humne pehle code pe .nunique lgaya tha then zx is converted in series and to convert it into dataframe we use rset_index.   
zx.rename(columns={'Customer Name':'No. Of Customers'},inplace=True)
zy = df.groupby(['State'])['Sales'].sum()
zxy = pd.merge(zx,zy,how='left',on=['State']).reset_index()# or yha reset_index islie lgaya bcoz plotly me graph tbhi kaam krta hai jb uske indexes ko reset kr dia jata hai...
zxy

In [ ]:
fig=px.scatter(zxy,'No. Of Customers','Sales',color='State',title='No. of Customers & Sales Co-Relation',
               hover_name='State',size='No. Of Customers')
fig.show()

Q-- What is Total share of States in genrating profit?

In [ ]:
wx = df.groupby(['State'])['Profit'].sum().reset_index()
wx

In [ ]:
fig= px.pie(wx,names=wx['State'],values=wx['Profit'],title='Profit Output By State')
fig.show()
# here names and values these 2 things r there which means - pie chart ko banane k lie 2 chije chahiye hoti hai, 1st=name and 2nd = value, - kis chij ka name tum vha show krna chah rhe ho....so here we write -wx dataframe ka state column ko hum as a name lenge and pie chart me kis chij ki values daalna chahte ho, to hum isme wx dataFrme k profit column ko as values choose krte hai.

# mene alag method se same pie chart niche b banaya hai.

In [ ]:
fig = px.pie(wx,'State','Profit',color='State',title='Profit output by state',hover_name='State')
fig.show()
# hover name is used to add extra information to graph when we put pointer on it.
# here instead of using names and values, we directly use columns name-- state and profit becoz it is by default that in first position- names will come and in second position values will come, it is like x and y axis. 

Q--. What is the share of each states in genrating sales for wallmart?

In [ ]:
 wy = df.groupby('State')['Sales'].sum().reset_index()
 wy

In [ ]:
fig = px.pie(wy,'State','Sales',title='Sales output by each state',hover_name='State')
fig.show()

California has the biggest share and genrates 63% of wallmart's revennue. Washington stand 2nd with contribution of 19.1%.

Q--. What is the share of each state in terms of footfall?    ...# footfall means how many customers  came into wallmart

In [ ]:
wz = df.groupby('State')['Customer Name'].nunique().reset_index()
wz.rename(columns={'Customer Name':'No. of unique Customers'},inplace=True)
wz

In [ ]:
fig = px.pie(wz,'State','No. of unique Customers',title= 'footfall per State',hover_name='State')
fig.show()

California is the 1st in the terms of footfall and has 61.5% of total footfall in wallmart.

In [ ]:
cd= df.groupby(['State', 'City'])['Customer Name'].nunique()

# Find the city with the highest number of customers in each state
best_cities = cd.groupby('State').nlargest()
best_cities
# .nlargest() method will not work in dataframe, it only works in series object....so if u write .reset_index() in 'cd' vaali line -it will gives u error. 

In [ ]:
cd= df.groupby(['State', 'City'])['Customer Name'].nunique()

# Find the city with the highest number of customers in each state
best_cities = cd.groupby('State').nlargest(1).reset_index(level=1, drop=True)
best_cities


In [ ]:
best_cities = df.groupby(['State', 'City'])['Customer Name'].nunique().reset_index()
idx = best_cities.groupby('State')['Customer Name'].idxmax()
best_cities.loc[idx, ['State', 'City']]


In [ ]:
best_cities = df.groupby(['State', 'City'])['Customer Name'].nunique().reset_index()
idx = best_cities.groupby('State')['Customer Name'].idxmax()
result = best_cities.loc[idx, ['State', 'City', 'Customer Name']].reset_index(drop=True)
result


In [ ]:
# i write this code, but this code is for to see how .size() and .nlargest() method works. 
#counts_by_state = df.groupby(['State', 'City']).size().groupby('State').nlargest(1)

#print(counts_by_state) 

In [ ]:
#m['Customers'].nlargest(10)

Q--.In each State which is the best performing city in term of no. of customers?

In [ ]:
xs = df.groupby(['State','City'])['Customer Name'].nunique().reset_index()
es = xs.groupby(['State'])['Customer Name'].idxmax()
result = xs.loc[es]
result

In [ ]:
fig = px.bar(result,'State','Customer Name',color='State',title='"Best Performing Cities by Number of Customers per State"')
fig.show()

In [ ]:
# now find top 10 customers which generated highest profit

In [ ]:

cv=df.groupby('Customer Name')['Profit'].sum()
yv=df.groupby(['Customer Name'])['Quantity'].sum()
iv=df.groupby(['Customer Name'])['Sales'].sum()
qv = pd.merge(pd.merge(cv,yv,how='left',on='Customer Name'),iv,how='left',on='Customer Name')
qv
# hum keval customer name par merge or groupby nhi kr skte becoz koi baar aisa hota hai ki alag alag states me same naam or caste k 2 persons ho skte hai...islie ye code shi nhi hai 2nd vaala code shi hai.

In [ ]:
cv=df.groupby(['State','Customer Name'])['Profit'].sum()
yv=df.groupby(['State','Customer Name'])['Quantity'].sum()
iv=df.groupby(['State','Customer Name'])['Sales'].sum()
qv=pd.merge(pd.merge(cv,yv,how='left',on=['State','Customer Name']),iv,how='left',on=['State','Customer Name'])
#qv.sort_values(by='Customer Name')  -- to aapko ye pta chalega ki ek naam k 2 persons b hai alag alag states me.
qv
#  in dono code ko jb aap run kroge to 1st set of code me 686 rows dega and 2nd set of code me arnd 2200 rows hongi or jb merge hme on=customer name pe hi krna hai to state column ho ya na ho usse to koi farak pdna nhi chahiye... but this 2nd code is right

In [ ]:
cv = df.groupby(['State','Customer Name'])['Profit'].sum().reset_index()# yha pe reset_index()-toconvert into dataframe

cv = cv.sort_values(by='Profit',ascending=False).reset_index().head(10)
cv
# you can get top 10 customers by this code also and by nlargest also which is written in next code.

In [ ]:
xx = qv.nlargest(10,'Profit')
xx.reset_index(inplace=True) # or assign reset_index() to a variable
xx

 Q--What 10 Customers have genrated us the most Profit?

In [ ]:
fig=px.bar(xx,'Customer Name','Profit',color='Customer Name',title= 'Top 10 most profitable customer')
fig.show()

Q--In each state which is the top customer in terms of profit?

In [ ]:
''' imp information'''
#top_cs = df.groupby(['State','Sales'])['Profit'].sum().reset_index()
#top_cs=top_cs.groupby(['State']).idxmax()
#top_cs
# idxmax() method tbhi lgega jb kewal ek column ho jisme numerical value ho...in this there is 2 numerical values therefore it gives us wrong answer

In [ ]:
top_cs = df.groupby(['State','Customer Name'])['Profit'].sum().reset_index()
max_idx=top_cs.groupby(['State'])['Profit'].idxmax()
max_value=top_cs.loc[max_idx].reset_index()
max_value

Q--In each state which is the top customer in terms of profit?

In [ ]:
fig= px.bar(max_value,'State','Profit',color='State',hover_name='Customer Name')
fig.show()
# hover_name-- gives you or adds  extra information on graphs, when you click on the ars of the data

Q--. Which Top 10 customers has caused the most losses to us?

In [ ]:
most_losses = df.groupby(['State','City','Customer Name'])['Profit'].sum().reset_index()
most_losses = most_losses.sort_values(by='Profit', ascending=True).head(10).reset_index()
most_losses = most_losses.rename(columns={'Profit':'Loss'})
most_losses
# In series 'by' parameter not works in sort_values, so you should apply reset_index (in first line)by this it converts into dataframe then you can able to apply sort_values.


In [ ]:
fig=px.bar(most_losses,'Customer Name','Loss',color='Customer Name',title= 'Top 10 losses in Wallmart',hover_name='City')
fig.show()

Q-- In each state which is the Top customer in terms of losess

In [ ]:
dp = df.groupby(['State','City','Customer Name'])['Profit'].sum().reset_index()

sp=dp.sort_values(by='Profit',ascending = True)
pdd = sp.groupby('State')['Profit'].idxmin()
result_2 = sp.loc[pdd]
result_2= result_2.rename(columns={'Profit':'Loss'})
result_2
# as we use idxmax- to get index of max value, - and -same we can use idxmin- to get index of min value 

In [ ]:
fig = px.bar(result_2,'State','Loss',color='State',hover_name='Customer Name')
fig.show()

Monthly Sales over months of all years ?

In [ ]:
monthly_sales = df.groupby(pd.Grouper(key='Order Date', freq='M'))['Sales'].sum()
#monthly_sales_growth = monthly_sales.pct_change()
monthly_sales_growth=pd.DataFrame(monthly_sales_growth)
monthly_sales_growth
# this code if u want sales in percentage and ask this code from chat gpt  it will explain u more clearly.

In [ ]:
monthly_sales = df.groupby(pd.Grouper(key='Order Date', freq='M'))['Sales'].sum()
#monthly_sales = monthly_sales.reset_index()
#monthly_sales = monthly_sales.rename(columns={'Sales': 'Total Sales'})
monthly_sales
# and this code gives u total sales of each month.
